File: backend/internal/maps/maps.go
package maps

import (
	"crypto/md5"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sort"
	"strings"
	"time"

	"4ever/backend/internal/config"
	"4ever/backend/internal/httputil"
	"github.com/gin-gonic/gin"
)

const tencentSuggestionURL = "https://apis.map.qq.com/ws/place/v1/suggestion"

type Handler struct {
	Settings config.Settings
}

type CityResult struct {
	ID     string  `json:"id"`
	Name   string  `json:"name"`
	Region string  `json:"region"`
	Lat    float64 `json:"lat"`
	Lon    float64 `json:"lon"`
}

func Register(group *gin.RouterGroup, h Handler) {
	r := group.Group("/maps")
	r.GET("/tencent/config", h.Config)
	r.GET("/tencent/city-search", h.CitySearch)
}

func (h Handler) Config(c *gin.Context) {
	if h.Settings.TencentMapKey == "" {
		httputil.Error(c, http.StatusServiceUnavailable, "Tencent map key is not configured.")
		return
	}
	c.JSON(http.StatusOK, gin.H{"map_key": h.Settings.TencentMapKey})
}

func (h Handler) CitySearch(c *gin.Context) {
	rawQuery, exists := c.GetQuery("q")
	if !exists {
		httputil.Error(c, http.StatusUnprocessableEntity, "q is required.")
		return
	}
	if len([]rune(rawQuery)) > 80 {
		httputil.Error(c, http.StatusUnprocessableEntity, "q must be 80 characters or fewer.")
		return
	}
	if h.Settings.TencentMapKey == "" {
		httputil.Error(c, http.StatusServiceUnavailable, "Tencent map key is not configured.")
		return
	}
	keyword := strings.TrimSpace(rawQuery)
	if keyword == "" {
		c.JSON(http.StatusOK, gin.H{"results": []CityResult{}})
		return
	}
	params := map[string]string{
		"keyword":    keyword,
		"key":        h.Settings.TencentMapKey,
		"region":     "中国",
		"region_fix": "0",
		"policy":     "1",
		"page_size":  "10",
	}
	if h.Settings.TencentMapSignatureKey != "" {
		params["sig"] = tencentSignature(params, h.Settings.TencentMapSignatureKey)
	}
	values := url.Values{}
	for key, value := range params {
		values.Set(key, value)
	}
	client := http.Client{Timeout: 8 * time.Second}
	resp, err := client.Get(tencentSuggestionURL + "?" + values.Encode())
	if err != nil {
		httputil.Error(c, http.StatusBadGateway, "Tencent city search failed.")
		return
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	if resp.StatusCode >= 400 {
		httputil.Error(c, http.StatusBadGateway, "Tencent city search failed.")
		return
	}
	var payload map[string]any
	if err := json.Unmarshal(body, &payload); err != nil {
		httputil.Error(c, http.StatusBadGateway, "Tencent city search failed.")
		return
	}
	if fmt.Sprint(payload["status"]) != "0" && payload["status"] != float64(0) {
		message := fmt.Sprint(payload["message"])
		if message == "" || message == "<nil>" {
			message = "Tencent city search failed."
		}
		httputil.Error(c, http.StatusBadGateway, message)
		return
	}
	results := []CityResult{}
	seen := map[string]bool{}
	items, _ := payload["data"].([]any)
	for _, item := range items {
		row, ok := item.(map[string]any)
		if !ok {
			continue
		}
		location, _ := row["location"].(map[string]any)
		lat, ok1 := numberValue(location["lat"])
		lon, ok2 := numberValue(location["lng"])
		title := strings.TrimSpace(fmt.Sprint(row["title"]))
		if title == "" || !ok1 || !ok2 {
			continue
		}
		parts := []string{}
		used := map[string]bool{}
		for _, key := range []string{"province", "city", "district"} {
			value := strings.TrimSpace(fmt.Sprint(row[key]))
			if value != "" && value != "<nil>" && !used[value] {
				parts = append(parts, value)
				used[value] = true
			}
		}
		region := strings.Join(parts, " · ")
		if region == "" {
			region = strings.TrimSpace(fmt.Sprint(row["address"]))
		}
		if region == "" || region == "<nil>" {
			region = "中国"
		}
		id := fmt.Sprintf("tencent:%s:%.6f:%.6f", title, lat, lon)
		if seen[id] {
			continue
		}
		seen[id] = true
		results = append(results, CityResult{ID: id, Name: title, Region: region, Lat: lat, Lon: lon})
	}
	c.JSON(http.StatusOK, gin.H{"results": results})
}

func tencentSignature(params map[string]string, key string) string {
	keys := make([]string, 0, len(params))
	for name := range params {
		keys = append(keys, name)
	}
	sort.Strings(keys)
	parts := make([]string, 0, len(keys))
	for _, name := range keys {
		parts = append(parts, name+"="+params[name])
	}
	sum := md5.Sum([]byte("/ws/place/v1/suggestion?" + strings.Join(parts, "&") + key))
	return hex.EncodeToString(sum[:])
}

func numberValue(value any) (float64, bool) {
	switch typed := value.(type) {
	case float64:
		return typed, true
	case int:
		return float64(typed), true
	default:
		return 0, false
	}
}

File: python_backend/app/maps.py
from __future__ import annotations

import hashlib
import json
from typing import Any
from urllib.parse import urlencode

from fastapi import APIRouter, HTTPException
import httpx

from app.config import Settings


TENCENT_SUGGESTION_URL = "https://apis.map.qq.com/ws/place/v1/suggestion"


def router(settings: Settings) -> APIRouter:
    api = APIRouter(prefix="/api/maps")

    @api.get("/tencent/config")
    def tencent_config() -> dict[str, str]:
        if not settings.tencent_map_key:
            raise HTTPException(status_code=503, detail="Tencent map key is not configured.")
        return {"map_key": settings.tencent_map_key}

    @api.get("/tencent/city-search")
    async def city_search(q: str | None = None) -> dict[str, Any]:
        if q is None:
            raise HTTPException(status_code=422, detail="q is required.")
        if len(q) > 80:
            raise HTTPException(status_code=422, detail="q must be 80 characters or fewer.")
        if not settings.tencent_map_key:
            raise HTTPException(status_code=503, detail="Tencent map key is not configured.")
        keyword = q.strip()
        if not keyword:
            return {"results": []}
        params = {
            "keyword": keyword,
            "key": settings.tencent_map_key,
            "region": "中国",
            "region_fix": "0",
            "policy": "1",
            "page_size": "10",
        }
        if settings.tencent_map_signature_key:
            params["sig"] = tencent_signature(params, settings.tencent_map_signature_key)
        try:
            async with httpx.AsyncClient(timeout=8) as client:
                response = await client.get(TENCENT_SUGGESTION_URL + "?" + urlencode(params))
        except httpx.HTTPError as error:
            raise HTTPException(status_code=502, detail="Tencent city search failed.") from error
        if response.status_code >= 400:
            raise HTTPException(status_code=502, detail="Tencent city search failed.")
        try:
            payload = response.json()
        except json.JSONDecodeError as error:
            raise HTTPException(status_code=502, detail="Tencent city search failed.") from error
        if str(payload.get("status")) not in {"0", "0.0"} and payload.get("status") != 0:
            raise HTTPException(status_code=502, detail=str(payload.get("message") or "Tencent city search failed."))
        return {"results": parse_city_results(payload)}

    return api


def tencent_signature(params: dict[str, str], key: str) -> str:
    parts = "&".join(f"{name}={params[name]}" for name in sorted(params))
    return hashlib.md5(("/ws/place/v1/suggestion?" + parts + key).encode("utf-8")).hexdigest()


def parse_city_results(payload: dict[str, Any]) -> list[dict[str, Any]]:
    results: list[dict[str, Any]] = []
    seen: set[str] = set()
    for item in payload.get("data") if isinstance(payload.get("data"), list) else []:
        if not isinstance(item, dict):
            continue
        location = item.get("location") if isinstance(item.get("location"), dict) else {}
        lat = location.get("lat")
        lon = location.get("lng")
        title = str(item.get("title") or "").strip()
        if not title or not isinstance(lat, (int, float)) or not isinstance(lon, (int, float)):
            continue
        region_parts: list[str] = []
        used: set[str] = set()
        for key in ("province", "city", "district"):
            value = str(item.get(key) or "").strip()
            if value and value not in used:
                region_parts.append(value)
                used.add(value)
        region = " · ".join(region_parts) or str(item.get("address") or "").strip() or "中国"
        city_id = f"tencent:{title}:{float(lat):.6f}:{float(lon):.6f}"
        if city_id in seen:
            continue
        seen.add(city_id)
        results.append({"id": city_id, "name": title, "region": region, "lat": float(lat), "lon": float(lon)})
    return results
